x-pack/solutions/observability/plugins/streams_app/public/hooks/queries/use_async_sample.tsx

```diff
@@ -0,0 +1,318 @@
+import {
+  Condition,
+  RecursiveRecord,
+  WiredStreamGetResponse,
+  conditionToQueryDsl,
+  getFields,
+} from '@kbn/streams-schema';
+import useToggle from 'react-use/lib/useToggle';
+import { MappingRuntimeField, MappingRuntimeFields } from '@elastic/elasticsearch/lib/api/types';
+import { filter, switchMap } from 'rxjs';
+import { isRunningResponse } from '@kbn/data-plugin/common';
+import { useKibana } from '../use_kibana';
+import { emptyEqualsToAlways } from '../../util/condition';
+
+interface Options {
+  condition?: Condition;
+  start?: number;
+  end?: number;
+  size?: number;
+  streamDefinition: WiredStreamGetResponse;
+}
+
+export const useAsyncSample = (options: Options) => {
+  const {
+    dependencies: {
+      start: { data },
+    },
+  } = useKibana();
+
+  // Documents
+  const [isLoadingDocuments, toggleIsLoadingDocuments] = useToggle(false);
+  const [documentsError, setDocumentsError] = useState();
+  const [documents, setDocuments] = useState<RecursiveRecord[]>([]);
+
+  // Document counts / percentage
+  const [isLoadingDocumentCounts, toggleIsLoadingDocumentCounts] = useToggle(false);
+  const [documentCountsError, setDocumentCountsError] = useState();
+  const [approximateMatchingPercentage, setApproximateMatchingPercentage] = useState<
+    string | undefined
+  >();
+
+  const [refreshId, setRefreshId] = useState(0);
+
+  const convertedCondition = useMemo(() => {
+    const condition = options.condition ? emptyEqualsToAlways(options.condition) : undefined;
+    return condition && 'always' in condition ? undefined : condition;
+  }, [options.condition]);
+
+  useEffect(() => {
+    if (!options.start || !options.end) {
+      setDocuments([]);
+      setApproximateMatchingPercentage(undefined);
+      return;
+    }
+
+    const runtimeMappings = getRuntimeMappings(options.streamDefinition, convertedCondition);
+
+    // Documents
+    toggleIsLoadingDocuments(true);
+    setDocuments([]);
+    const documentSubscription = data.search
+      .search({
+        params: {
+          index: options.streamDefinition.stream.name,
+          body: getDocumentsSearchBody(options, runtimeMappings, convertedCondition),
+        },
+      })
+      .subscribe({
+        next: (result) => {
+          if (!isRunningResponse(result)) {
+            toggleIsLoadingDocuments(false);
+          }
+
+          if (result.rawResponse.hits?.hits) {
+            setDocuments((prev) => result.rawResponse.hits.hits.map((hit) => hit._source));
+          }
+        },
+        error: (e) => {
+          setDocumentsError(e);
+          toggleIsLoadingDocuments(false);
+        },
+      });
+
+    toggleIsLoadingDocumentCounts(true);
+    setApproximateMatchingPercentage(undefined);
+    const documentCountsSubscription = data.search
+      .search({
+        params: {
+          index: options.streamDefinition.stream.name,
+          body: getDocumentCountForSampleRateSearchBody(options),
```

Comment:
Currently we are re-fetching the full count for probability calculation every time - should we cache this based on the current body? This way we can remove another roundtrip from the response